src/Microsoft.CodeQuality.Analyzers/UnitTests/ApiDesignGuidelines/EnumShouldNotHaveDuplicatedValuesTests.cs

```diff
@@ -0,0 +1,411 @@
+Public Enum MyEnum
+    None = &H0
+    Value1 = &H1
+    Value2 = &H1
+    Value3 = &H4
+End Enum",
+                GetBasicResultAt(5, 5));
+        }
+
+        [Fact]
+        public async Task EnumImplicitDuplication_Diagnostic()
+        {
+            await VerifyCS.VerifyAnalyzerAsync(@"
+public enum MyEnum
+{
+    Value1 = 1,
+    Value2,
+    Value3 = 2,
+}",
+                GetCSharpResultAt(6, 5));
+
+            await VerifyVB.VerifyAnalyzerAsync(@"
+Public Enum MyEnum
+    Value1 = 1
+    Value2
+    Value3 = 2
+End Enum",
+                GetBasicResultAt(5, 5));
+        }
+
+        [Fact]
+        public async Task EnumValueReference_NoDiagnostic()
+        {
+            await VerifyCS.VerifyAnalyzerAsync(@"
+public enum MyEnum
+{
+    Value1 = 1,
+    Value2 = 2,
+    Value3 = Value1,
+}");
+
+            await VerifyVB.VerifyAnalyzerAsync(@"
+Public Enum MyEnum
+    Value1 = 1
+    Value2 = 2
+    Value3 = Value1
+End Enum");
+        }
+
+        [Fact]
+        public async Task EnumDuplicatedBitwiseValueReference_Diagnostic()
+        {
+            await VerifyCS.VerifyAnalyzerAsync(@"
+public enum MyEnum
+{
+    Value1 = 1,
+    Value2 = 2,
+    Value3 = Value1 | Value1,
+}",
+                GetCSharpResultAt(6, 5),
+                GetCSharpBitwiseResultAt(6, 23));
+
+            await VerifyVB.VerifyAnalyzerAsync(@"
+Public Enum MyEnum
+    Value1 = 1
+    Value2 = 2
+    Value3 = Value1 Or Value1
+End Enum",
+                GetBasicResultAt(5, 5),
+                GetBasicBitwiseResultAt(5, 24));
+        }
+
+        [Fact]
+        public async Task EnumDuplicatedBitwiseValueReferenceComplex_Diagnostic()
+        {
+            await VerifyCS.VerifyAnalyzerAsync(@"
+public enum MyEnum
+{
+    Value1 = 1,
+    Value2 = 2,
+    Value3 = 3,
+    Value4 = Value1 | Value2 | Value1,
+}",
+                GetCSharpResultAt(7, 5),
+                GetCSharpBitwiseResultAt(7, 32));
+
+            await VerifyVB.VerifyAnalyzerAsync(@"
+Public Enum MyEnum
+    Value1 = 1
+    Value2 = 2
+    Value3 = 3
+    Value4 = Value1 Or Value2 Or Value1
+End Enum",
+                GetBasicResultAt(6, 5),
+                GetBasicBitwiseResultAt(6, 34));
+        }
+
+        [Fact]
+        public async Task EnumDuplicatedBitwiseValueReferenceAndValue_Diagnostic()
+        {
+            await VerifyCS.VerifyAnalyzerAsync(@"
+public enum MyEnum
+{
+    Value1 = 1,
+    Value2 = 2,
+    Value3 = Value1 | 1,
+}",
+                GetCSharpResultAt(6, 5));
+
+            await VerifyVB.VerifyAnalyzerAsync(@"
+Public Enum MyEnum
+    Value1 = 1
+    Value2 = 2
+    Value3 = Value1 Or 1
+End Enum",
+                GetBasicResultAt(5, 5));
+        }
+
+        [Fact]
+        public async Task EnumDuplicatedValueMaths_Diagnostic()
+        {
+            await VerifyCS.VerifyAnalyzerAsync(@"
+public enum MyEnum
+{
+    Value1 = 1,
+    Value2 = 2 - 1,
+    Value3 = Value1 * Value2,
+}",
+                GetCSharpResultAt(5, 5),
```

Comment:
I think it feels worthwhile to mention the duplicated value and the field names with the duplicated value in the diagnostic message. Would that be easy to implement?

Reply:
I am not entirely sure to get what you would like.

Taking this code as an example:

```csharp
public enum MyEnum
{
    Value1 = 1,
    Value2 = 2 - 1,
    Value3 = Value1 * Value2,
}
```

We currently report 1 issue on `Value2` and 1 on `Value3`.

Are you suggesting to

1. Tweak the message to say that `Value2` duplicates `Value1`?

2. Report only once on `Value1` saying that `Value2` and `Value3` duplicates this field?

Reply:
Reporting once on each field seems reasonable. May be the diagnostic on `Value1` can say that Field `Value1` has the same constant value `x` as field `Value2`? Not required to implement, just a suggestion.